src/upload/form/mod.rs

```diff
@@ -53,38 +58,88 @@ pub fn build(
                 plain_texts,
             )?;
 
-            build_form(&assets, session_config)
-        }
-        TargetType::JavaScript => {
-            log::info!("JavaScript project detected. Publishing...");
-            let build_dir = target.build_dir()?;
-            let package = Package::new(&build_dir)?;
-
-            let script_path = package.main(&build_dir)?;
-
-            let assets = ProjectAssets::new(
-                script_path,
-                wasm_modules,
-                kv_namespaces.to_vec(),
-                text_blobs,
-                plain_texts,
-            )?;
-
-            build_form(&assets, session_config)
+            service_worker::build_form(&assets, session_config)
         }
+        TargetType::JavaScript => match &target.builder_config {
+            Some(config) => match &config.upload_format {
+                ScriptFormat::ServiceWorker => {
+                    log::info!("Plain JavaScript project detected. Publishing...");
+                    let package_dir = target.package_dir()?;
+                    let package = Package::new(&package_dir)?;
+                    let script_path = package.main(&package_dir)?;
+
+                    let assets = ServiceWorkerAssets::new(
+                        script_path,
+                        wasm_modules,
+                        kv_namespaces.to_vec(),
+                        text_blobs,
+                        plain_texts,
+                    )?;
+
+                    service_worker::build_form(&assets, session_config)
+                }
+                ScriptFormat::Modules => {
+                    let package_dir = target.package_dir()?;
+                    let package = Package::new(&package_dir)?;
+                    let main_module = package.module(&package_dir)?;
+                    let main_module_name = filename_from_path(&main_module)
+                        .ok_or_else(|| failure::err_msg("filename required for main module"))?;
+
+                    let modules_iter = WalkBuilder::new(config.build_dir.clone())
+                        .standard_filters(false)
+                        .build();
+
+                    let mut modules: Vec<Module> = vec![];
+
+                    for entry in modules_iter {
+                        let entry = entry?;
+                        let path = entry.path();
+                        if path.is_file() {
+                            modules.push(Module::new(path.to_owned())?);
+                        }
+                    }
+
+                    let assets = ModulesAssets::new(
+                        main_module_name,
+                        modules,
+                        kv_namespaces.to_vec(),
+                        plain_texts,
+                    )?;
+
+                    modules_worker::build_form(&assets, session_config)
+                }
+            },
+            None => {
+                log::info!("Plain JavaScript project detected. Publishing...");
```

Comment:
this is duplication of the ServiceWorker case above, right?

Reply:
The service worker case above is effectively the same by herre, but differs earlier along in the process since it's possible to use the service_worker upload type with custom builds. I'd like to keep this support, as looking at the issue tracker it will unblock quite a few folks.